exp: add rename command

BradyJ27 opened the pull request:
* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

https://github.com/iterative/dvc.org/issues/4793 and Open PR: https://github.com/iterative/dvc.org/pull/4805

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

Adds rename command for renaming finished experiments. Renaming queued experiments is not seen as an essential at this time. This PR is barebones and like will go back into development, but just would like some maintainers to take a look so I can make a list of improvements. 

Fixes: #6667

BradyJ27 commented on dvc/commands/experiments/rename.py:
```diff
@@ -0,0 +1,83 @@
+        ) ^ bool(self.args.experiment):
+            raise InvalidArgumentError(
+                "Either provide an `experiment` argument, or use the "
+                "`--rev` or `--all-commits` or `--queue` flag."
```
Remove "all-commits" and "--queue"

BradyJ27 commented on dvc/commands/experiments/rename.py:
```diff
@@ -0,0 +1,83 @@
+        "rename",
+        parents=[parent_parser],
+        description=append_doc_link(EXPERIMENTS_REMOVE_HELP, "exp/remove"),
+        help=EXPERIMENTS_REMOVE_HELP,
```
Change this to `EXPERIMENTS_RENAME_HELP`. Also, create an issue in dvc.org at some point

BradyJ27 commented on dvc/commands/experiments/rename.py:
```diff
@@ -0,0 +1,83 @@
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+    )
+    remove_group = experiments_rename_parser.add_mutually_exclusive_group()
+    remove_group.add_argument(
```
`rename_group`

BradyJ27 commented on dvc/repo/experiments/__init__.py:
```diff
@@ -398,6 +398,11 @@ def ls(self, *args, **kwargs):
+        from dvc.repo.experiments.rename import rename
+
+        return rename(self.repo, *args, **kwargs)
+
```
Move this below `remove`

shcheklein commented:
@iterative/dvc can we please review this?

BradyJ27 commented:
I've opened a docs PR as well: https://github.com/iterative/dvc.org/pull/4805

daavoo commented on dvc/commands/experiments/rename.py:
```diff
@@ -0,0 +1,82 @@
+        f"Rename experiments derived from the specified `<commit>` as "
+        f"baseline{default_msg}."
+    )
+    experiments_rename_parser.add_argument(
```
I am not sure about the `rev` usage. It feels convenient but I would prefer to lie on the side of simplicity.

It would be also simple enough to pipe `dvc exp list` and this command to achieve the same and an example in the docs could be enough.

BradyJ27 replied:
Are you suggesting removal of the --rev flag all together? I agree, I believe it also came up in the docs review, that it is a little weird/unclear.

daavoo replied:
Yes, I would prefer to just keep the simple positional arguments

dberenbaum replied:
Sorry, I commented in the docs PR without looking at the core PR here closely. Agree with @daavoo, let's drop `--rev` please for now.

BradyJ27 replied:
Removed --rev flag

daavoo commented on tests/func/experiments/test_rename.py:
Would be good to also test a no-op (name not found)

BradyJ27 replied:
Sure, just added that.

daavoo commented on dvc/repo/experiments/rename.py:
```diff
@@ -0,0 +1,58 @@
+    if exp_name:
+        results: Dict[
+            str, ExpRefAndQueueEntry
+        ] = celery_queue.get_ref_and_entry_by_names(exp_name, git_remote)
```
I think we could just use `from dvc.repo.experiments.utils import resolve_name` instead of this method

BradyJ27 replied:
You're right. I've updated this

daavoo reviewed (changes requested):
Thank you so much for the contribution @BradyJ27 and also for putting the docs 🙏 

This feels like a useful addition to me.

I have 2 major considerations (`--rev` and internal `get_ref_and_entry_by_names` usage) but it is already in good shape.

BradyJ27 commented:
Thanks for the reviews @daavoo @dberenbaum. I have updated the PR and added all requested changes. I also updated the docs PR to reflect these changes.

daavoo commented on dvc/commands/experiments/rename.py:
```diff
@@ -0,0 +1,64 @@
+
+
+class CmdExperimentsRename(CmdBase):
+    def check_arguments(self):
+        if not bool(self.args.name) | bool(self.args.experiment):
+            raise InvalidArgumentError(
+                "Please provide an `experiment` and `name` argument."
+            )
+
+    def run(self):
+        from dvc.utils import humanize
+
+        self.check_arguments()
```
This should be handled by argparse , I think the code would never run as it would fail when parsing args.

```suggestion
    def run(self):
        from dvc.utils import humanize
```

daavoo commented on tests/func/experiments/test_rename.py:
What happens when the provided name already exists?
Can we add a test for it?

BradyJ27 replied:
This used to be handled by just renaming to the same name, however, I've changed it to now return None, which returns to the user "No experiments to be renamed"

daavoo reviewed (changes requested):
Thanks for quickly addressing the changes @BradyJ27 .

First, the CI is failing from the Python 3.10 only syntax:

https://github.com/iterative/dvc/actions/runs/6055886949/job/16435317074?pr=9812#step:5:323

I left some additional comments from things I found in this review (removing the code should fix the CI) sorry for not bringing them in the previous one.

It would also be good to add a test like the existing ones in https://github.com/iterative/dvc/blob/main/tests/unit/command/test_experiments.py.

BradyJ27 commented on tests/func/experiments/test_rename.py:
```diff
@@ -0,0 +1,10 @@
+
+def test_rename_experiment_by_name(scm, dvc, exp_stage):
+    dvc.experiments.run(exp_stage.addressing, name="test-name", params=["foo=1"])
+    assert dvc.experiments.rename("new-name", "test-name") == ["test-name"]
```
Ok sure, updated the tests

BradyJ27 commented:
> Thanks for quickly addressing the changes @BradyJ27 .
> 
> First, the CI is failing from the Python 3.10 only syntax:
> 
> https://github.com/iterative/dvc/actions/runs/6055886949/job/16435317074?pr=9812#step:5:323
> 
> I left some additional comments from things I found in this review (removing the code should fix the CI) sorry for not bringing them in the previous one.
> 
> It would also be good to add a test like the existing ones in https://github.com/iterative/dvc/blob/main/tests/unit/command/test_experiments.py.
> 

Sorry @daavoo , I went on a bit of a hiatus. I've resolved all of your comments and added a few tests in `test_experiments`

shcheklein commented on dvc/repo/experiments/rename.py:
```diff
@@ -0,0 +1,60 @@
+def _rename_exp(scm: "Git", ref_info: "ExpRefInfo", new_name: str):
+    rev = scm.get_ref(str(ref_info))
+    scm.remove_ref(str(ref_info))
+    ref_info.name = new_name
```
@BradyJ27 @pmrowla qq - do we allow any experiment name that is a valid ref name? (can it include `/` for example? that might start breaking some code downstream, at least we need to test it)

BradyJ27 replied:
I'm unsure how this will behave, I did not test when developing. I can test this and then get back based on the results. Are we thinking that we should just restrict user input to solve this?

pmrowla replied:
@shcheklein experiment names are allowed to use any valid git ref character except for `/`
